functions/src/index.ts

```diff
@@ -46,163 +45,194 @@ const genericErrorHandler = (err: any, req: express.Request, res: express.Respon
-            }
-            // context_id is optional so don't test for it
-            // TODO: context_id is still "class_hash" rigse jwt_controller.rb - should it be class_hash here?
-            req.claims = claims;
-            next();
-          }
-        })
-      }
-    })
-    .catch(error => res.error(403, error))
+const authUsingJWTOrReadWriteToken = (req: express.Request): JWTClaims | string => {
+  const token = getToken(req);
+  if (!token) {
+    throw new Error("Missing token in headers, query or cookie");
+  } else if (token.startsWith(ReadWriteTokenPrefix)) {
+    return token;
+  } else {
+    return getClaimsFromJWT(token);
+  }
 };
 
 app.use(customHandlers);
 app.use(genericErrorHandler);
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 app.use(cors());
-app.use(checkAuth);
-
-const checkEnv = (req: express.Request, res: express.Response, next: express.NextFunction) => {
-  const { env } = req.query;
-  if (!env) {
-    res.error(403, "Missing env query parameter!");
-  }
-  else {
-    req.env = env as string;
-    next();
-  }
-};
 app.use(checkEnv);
 
 app.get('/api/v1/resources', (req, res) => {
-  return BaseResourceObject.FindAll(db, req.env, req.claims, req.query)
-    .then(resources => resources.map((resource) => resource.apiResult()))
-    .then(apiResults => res.success(apiResults))
-    .catch(error => res.error(400, error))
-})
+  try {
+    const claims = authUsingJWT(req);
+    BaseResourceObject.FindAll(db, req.env, claims, req.query)
+      .then(resources => resources.map((resource) => resource.apiResult()))
+      .then(apiResults => res.success(apiResults))
+      .catch(error => res.error(400, error))
+  } catch (error) {
+    res.error(403, error)
+  }
+});
 
 app.get('/api/v1/resources/:id', (req, res) => {
-  return BaseResourceObject.Find(db, req.env, req.params.id)
+  BaseResourceObject.Find(db, req.env, req.params.id)
     .then(resource => res.success(resource.apiResult()))
     .catch(error => res.error(404, error))
-})
+});
 
 app.post('/api/v1/resources', (req, res) => {
-  return BaseResourceObject.Create(db, req.env, req.claims, req.body)
-    .then(resource => res.success(resource.apiResult(), 201))
-    .catch(error => res.error(400, error))
-})
+  // Note that this is the only route where auth is not called directly. Resource can be created anonymously.
+  const token = getToken(req);
+  if (token) {
+    try {
+      const claims = getClaimsFromJWT(token);
+      BaseResourceObject.Create(db, req.env, claims, req.body)
+        .then(resource => res.success(resource.apiResult(), 201))
+        .catch(error => res.error(400, error));
+    } catch (error) {
+      res.error(403, error);
+    }
+  } else {
+    // Anonymous usage based on readWriteToken. NOT RECOMMENDED apart from testing. readWriteTokens are meant to
+    // be used mostly/only to support converted Document Store documents.
```

Comment:
At this point I think we are going to stick with just using readWriteTokens for all shared documents. So the "not recommend" part of this should be removed.